### ISSUE TITLE
Model is Not Training Faster On Distributed Setup of Multi-Node Multi-GPU Compared to Single GPU

### ISSUE THREAD

**Issue by Ankush610** [GaParmar/img2img-turbo#154]
First of all the model does run on smaller batch size like 1,2,4 but even with batch size of 4 somehow single gpu outperforms the distributed setup in terms of training time. I did try to increase the batch size but any size greater than 4-6 gives OMM error so how does distributed setup suppose to be efficient?  

This is my Config: 
compute_environment: LOCAL_MACHINE
debug: true
distributed_type: MULTI_GPU
downcast_bf16: 'no'
enable_cpu_affinity: false
gpu_ids: all
machine_rank: 0
main_process_ip: 0   # will be overridden by --main_process_ip flag
main_process_port: 29900
main_training_function: main
mixed_precision: fp16
num_machines: 2
num_processes: 4
rdzv_backend: static
same_network: true
use_cpu: false 

slurm script :
 #!/bin/bash
#SBATCH --job-name=img2img-turbo-run
#SBATCH --nodes=2
#SBATCH --ntasks-per-node=1          # one launcher per node
#SBATCH --cpus-per-task=32
#SBATCH --gres=gpu:2
#SBATCH --partition=gpu
#SBATCH --output=logs_%j.out
#SBATCH --error=logs_%j.err
#SBATCH --time=24:00:00

# Activate environment
source path/to/miniconda3/bin/activate img2img-turbo

# Define and log node and distributed config
nodes_array=($(scontrol show hostnames $SLURM_JOB_NODELIST))
head_node=${nodes_array[0]}
head_node_ip=$(srun --nodes=1 --ntasks=1 -w "$head_node" hostname --ip-address)

echo "===================================="
echo "SLURM Job Information"
echo "===================================="
echo "Job Name: $SLURM_JOB_NAME"
echo "Number of Nodes: $SLURM_NNODES"
echo "GPUs per Node: $SLURM_GPUS_ON_NODE"
echo "Nodes Array: ${nodes_array[@]}"
echo "Head Node: $head_node"
echo "Head Node IP: $head_node_ip"
echo "===================================="

# Distributed env vars
export MASTER_ADDR=$head_node_ip
export MASTER_PORT=29900
export WORLD_SIZE=$(($SLURM_NNODES * $SLURM_GPUS_ON_NODE))
export RANK=$SLURM_PROCID
export OMP_NUM_THREADS=${SLURM_CPUS_PER_TASK}

# NCCL config
export NCCL_IB_DISABLE=0
export NCCL_SOCKET_IFNAME=ib0
export NCCL_DEBUG=INFO
export NCCL_PROTO=simple
export NCCL_ASYNC_ERROR_HANDLING=1
export NCCL_P2P_DISABLE=0
export NCCL_NET=IB 
export TORCHELASTIC_TIMEOUT=180
export NCCL_DEBUG=WARN
export WANDB_MODE=disabled
export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
export HF_HUB_OFFLINE=1 # load model offline
export TRANSFORMERS_OFFLINE=1

# HF/Accelerate log levels
export TRANSFORMERS_VERBOSITY=info
export HF_LOG_LEVEL=info

echo "===================================="
echo "Distributed Setup"
echo "===================================="
echo "MASTER_ADDR: $MASTER_ADDR"
echo "MASTER_PORT: $MASTER_PORT"
echo "WORLD_SIZE: $WORLD_SIZE"
echo "RANK: $RANK"
echo "OMP_NUM_THREADS: $OMP_NUM_THREADS"
echo "NCCL_IB_DISABLE: $NCCL_IB_DISABLE"
echo "NCCL_DEBUG: $NCCL_DEBUG"
echo "===================================="

# Launch job using Accelerate
echo "===================================="
echo "Starting Distributed Training with accelerate"
echo "===================================="

srun accelerate launch \
  --config_file accelerate_config.yaml \
  --main_process_ip $MASTER_ADDR \
  --main_process_port $MASTER_PORT \
  --num_processes $WORLD_SIZE \
  --num_machines $SLURM_NNODES \
  --machine_rank $RANK \
  --rdzv_backend c10d \
  src/train_cyclegan_turbo.py \
  --pretrained_model_name_or_path="stabilityai/sd-turbo" \
  --dataset_folder data/EYE \
  --train_img_prep "resize_286_randomcrop_256x256_hflip" \
  --val_img_prep "resize_256" \
  --learning_rate "1e-5" \
  --max_train_steps 2500 \
  --max_train_epochs 600 \
  --train_batch_size 4 \
  --gradient_accumulation_steps 1 \
  --tracker_project_name "cyclegan_turbo_optimized" \
  --enable_xformers_memory_efficient_attention \
  --validation_steps 250 \
  --lambda_gan 0.5 \
  --lambda_idt 1 \
  --lambda_cycle 1 \
  --dataloader_num_workers 32 \
  --output_dir /scratch/omjadhav/output/

echo "===================================="
echo "Training Completed"
echo "===================================="

**Comment by Aaron-116** [comment]
Hello, I have encountered the same issue, have you sloved it?

**Comment by Ankush610** [comment]
Hey, here’s a quick summary of what happened with the runs.

The turbo model (SD-Turbo based) failed mainly due to out-of-memory issues. That model includes several additional components like VAE, modified UNet blocks, and attention mechanisms, which significantly increase memory usage. Even with fp16 and xformers enabled, the per-GPU memory footprint was too high for me, which restricted the batch size to around 1–4. Because of that, distributed training became inefficient — each GPU was doing very little computation while still incurring the cost of gradient synchronization, leading to poor performance overall.

I then switched to the original implementation from pytorch-CycleGAN-and-pix2pix (https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/), which worked because it is much lighter. It has a simpler architecture, lower memory requirements, and allows a larger batch size (32), making the training setup stable.

However, even with the original model, distributed training is still slower than expected. This is not due to an error but rather the nature of CycleGAN itself. Each training step involves multiple sequential operations (two generators and two discriminators), which reduces parallel efficiency. Additionally, the computation per GPU is not very high compared to modern large-scale models, and there are frequent synchronization points between GPUs.

So even with a reasonable per-GPU batch size (~8), communication overhead and the sequential training structure limit scaling efficiency. As a result, the speedup from multiple GPUs is not linear.
